_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+
+<!--more-->
+
+### Delivery guarantees
+
+In distributed systems like Kafka, any component can fail independently and you have to deal with this fact.
+A broker may crash or a network failure may happen while the producer is sending a message to a topic.
+
+Depending on the configuration, you can have three different delivery semantics:
+
+- **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)
+- **at-least-once**: the producer retries but the broker can fail after writing but before sending the ack (duplicates risk)
+- **exactly-once**: even if the producer retries sending a message, it is written exactly once (idempotency)
+
+With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
+There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. financial institutions).
+
+To be fair, there is no such thing as exactly-once delivery in distributed systems (two generals problem), but we can fake it by implementing idempotent operations.
+Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.
+
+Since Kafka 3.0, the producer enables the strongest delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
+In addition to durability, this provides partition level message ordering and duplicates protection.
+
+When the idempotent producer is enabled, the broker registers the producer id (PID) and the epoch (producer session).
+A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
+The broker keeps track of the sequence number per producer and partition, periodically storing this information in `.snapshot` files.
+Whenever a new batch arrives, the broker checks if the last received number is equal to the first batch number plus one, then the batch is acknowledged, otherwise it is rejected.
+
+Unfortunately, the idempotent producer does not guarantee atomicity when you need to write to multiple partitions as a single unit of work.
+This is the typical scenario with many stream processing applications that run read-process-write cycles.
+In this case, you can use transactions API or Streams API to make these cycles atomic.
+The transaction API is low level and you need to ensure the right configuration and to commit offsets within the scope of the transaction.
+Instead, with the Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.
+
+### Considerations
+
+Each producer must configure its own static and unique `transactional.id` (TID) to enable the zombie fencing logic, which avoids message duplicates.
```

Comment:
What about linking `transactional.id` to the official docs?
https://kafka.apache.org/documentation/#producerconfigs_transactional.id

Reply:
Agree to both. This is a critical configuration and we don't want ambiguity.